eva-accession-pipeline/src/main/java/uk/ac/ebi/eva/accession/pipeline/steps/processors/ContigReplacerProcessor.java

```diff
@@ -0,0 +1,98 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package uk.ac.ebi.eva.accession.pipeline.steps.processors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.batch.item.ItemProcessor;
+import uk.ac.ebi.eva.accession.core.contig.ContigMapping;
+import uk.ac.ebi.eva.accession.core.contig.ContigSynonyms;
+import uk.ac.ebi.eva.commons.core.models.IVariant;
+import uk.ac.ebi.eva.commons.core.models.pipeline.Variant;
+import uk.ac.ebi.eva.commons.core.models.pipeline.VariantSourceEntry;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Converts the contig to it's GenBank synonym when possible. If the synonym can't be determined it keeps the contig as
+ * is
+ */
+public class ContigReplacerProcessor implements ItemProcessor<IVariant, IVariant> {
+
+    private static final Logger logger = LoggerFactory.getLogger(ContigReplacerProcessor.class);
+
+    private ContigMapping contigMapping;
+
+    private Set<String> processedContigs;
+
+    public ContigReplacerProcessor(ContigMapping contigMapping) {
+        this.contigMapping = contigMapping;
+        this.processedContigs = new HashSet<>();
+    }
+
+    @Override
+    public IVariant process(IVariant variant) throws Exception {
+        String contigName = variant.getChromosome();
+        ContigSynonyms contigSynonyms = contigMapping.getContigSynonyms(contigName);
+        boolean contigPresentInAssemblyReport = contigSynonyms != null;
+
+        if (!contigPresentInAssemblyReport) {
+            if (!processedContigs.contains(contigName)) {
+                logger.warn("Contig '" + variant.getChromosome() + "' was not found in the assembly report!");
+                processedContigs.add(contigName);
+            }
+            return variant;
+        } else {
+            return replaceContigWithSynonym(variant, contigSynonyms);
+        }
+    }
+
+    /**
+     * The conversion will be performed only if the 'relationship' column of the assembly report is '='
+     */
+    private IVariant replaceContigWithSynonym(IVariant variant, ContigSynonyms contigSynonyms) {
+        if (contigSynonyms.isIdenticalGenBankAndRefSeq()) {
+            Variant newVariant = new Variant(getContigSynonym(contigSynonyms), variant.getStart(), variant.getEnd(),
+                                             variant.getReference(), variant.getAlternate());
+            Collection<VariantSourceEntry> sourceEntries = variant.getSourceEntries().stream()
+                                                                  .map(VariantSourceEntry.class::cast)
+                                                                  .collect(Collectors.toList());
```

Comment:
ugh. This is so ugly but I can't think of a better thing to do. we don't currently need the source entries, but the factory will add them. we could:

- make the whole step to use `Variant` instead of `IVariant`.
- and/or use  `new VcfReader(new CoordinatesVcfLineMapper(), new File(inputParameters.getOutputVcf()));` and assert here that the sourceEntries is empty. after all we only care about chr, start, ref, alt. the project, taxonomy and assembly is provided as input parameter.
- or not copying the sourceEntries (dangerous or misleading to reuse this class?)
- or keep it as it is now?

any other idea?

Reply:
Although I don't like it a lot I would go for doing the cast (as it is now), my second option would be using `Variant` instead of `IVariant` and the cast wouldn't be needed.

I Will leave it with the cast unless someone would prefer another option.